packages/hw-app-ckb/src/Ckb.js

```diff
@@ -0,0 +1,180 @@
+
+  constructor(transport: Transport<*>, scrambleKey: string = "CKB") {
+    this.transport = transport;
+    transport.decorateAppAPIMethods(
+      this,
+      [
+        "getAppConfiguration",
+        "getWalletId",
+        "getWalletPublicKey",
+        "signTransaction"
+      ],
+      scrambleKey
+    );
+  }
+
+  /**
+   * get CKB address for a given BIP 32 path.
+   *
+   * @param path a path in BIP 32 format
+   * @return an object with a publicKey
+   * @example
+   * const result = await ckb.getWalletPublicKey("44'/144'/0'/0/0");
+   * const publicKey = result;
+   */
+  async getWalletPublicKey(path: string): Promise<string> {
+    const bipPath = BIPPath.fromString(path).toPathArray();
+
+    const cla = 0x80;
+    const ins = 0x02;
+    const p1 = 0x00;
+    const p2 = 0x00;
+    const data = Buffer.alloc(1 + bipPath.length * 4);
+
+    data.writeUInt8(bipPath.length, 0);
+    bipPath.forEach((segment, index) => {
+      data.writeUInt32BE(segment, 1 + index * 4);
+    });
+
+    const response = await this.transport.send(cla, ins, p1, p2, data);
+    const publicKeyLength = response[0];
+    return response.slice(1, 1 + publicKeyLength).toString("hex");
+  }
+
+  /**
+   * Sign a Nervos transaction with a given BIP 32 path
+   *
+   * @param path a path in BIP 32 format
+   * @param rawTxHex transaction to sign
+   * @param contextTransaction list of transaction context to use in parsing (optional)
+   * @return a signature as hex string
+   * @example
+   * const signature = await ckb.signTransaction("44'/144'/0'/0/0", "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");
+   */
+  async signTransaction(
+    path: string,
+    rawTxHex: string,
+    rawContextsTxHex: [string]
+  ): Promise<string> {
+    const bipPath = BIPPath.fromString(path).toPathArray();
+    const rawTx = Buffer.from(rawTxHex, "hex");
+
+    let rawPath = Buffer.alloc(1 + bipPath.length * 4);
+    rawPath.writeInt8(bipPath.length, 0);
+    bipPath.forEach((segment, index) => {
+      rawPath.writeUInt32BE(segment, 1 + index * 4);
+    });
+    await this.transport.send(0x80, 0x03, 0x00, 0x00, rawPath);
+
+    const maxApduSize = 230;
+
+    if (!rawContextsTxHex) rawContextsTxHex = [];
+
+    for (const rawContextTxHex of rawContextsTxHex) {
+      let rawContextTx =
+        rawContextTxHex !== null ? Buffer.from(rawTxHex, "hex") : null;
+      for (let i = 0; i < Math.floor(rawContextTx.length / maxApduSize); i++) {
+        let data = Buffer.alloc(maxApduSize);
+        for (let j = 0; j < maxApduSize; j++) {
+          let index = i * maxApduSize + j;
+          if (index >= rawContextTx.length) {
+            break;
+          }
+          rawContextTx.copy(data, 1, index);
```

Comment:
we seem to do byte per byte, it's not possible to do it in a chunk of byte with copy? apologies if i miss the big picture

Reply:
Fixed and the copy removed, even; it was not required.